python/xorbits/deploy/kubernetes/client.py

```diff
@@ -0,0 +1,506 @@
+            worker_cache_mem=self._worker_cache_mem,
+            min_cache_mem=self._worker_min_cache_men,
+            service_name=self._service_name,
+            service_port=self._worker_service_port,
+            pre_stop_command=self._pre_stop_command,
+            supervisor_web_port=self._web_port,
+            use_local_image=self._use_local_image,
+        )
+        workers_config.add_simple_envs(self._worker_extra_env)
+        workers_config.add_labels(self._worker_extra_labels)
+        workers_config.create_namespaced(self._api_client, self._namespace)
+
+    def _create_services(self):
+        self._create_supervisors()
+        self._create_workers()
+
+    def _wait_services_ready(self):
+        min_worker_num = int(self._min_worker_num or self._worker_num)
+        limits = [self._supervisor_num, min_worker_num]
+        selectors = [
+            "xorbits/service-type=" + XorbitsSupervisorsConfig.rc_name,
+            "xorbits/service-type=" + XorbitsWorkersConfig.rc_name,
+        ]
+        start_time = time.time()
+        logger.debug("Start waiting pods to be ready")
+        wait_services_ready(
+            selectors,
+            limits,
+            lambda sel: self._get_ready_pod_count(sel),
+            timeout=self._timeout,
+        )
+        logger.info("All service pods ready.")
+        if self._timeout is not None:  # pragma: no branch
+            self._timeout -= time.time() - start_time
+
+    def _create_ingress(self):
+        self._networking_api.create_namespaced_ingress(
+            self._namespace,
+            IngressConfig(
+                self._namespace,
+                self._ingress_name,
+                self._service_name,
+                self._web_port,
+                self._cluster_type,
+            ).build(),
+        )
+
+    def _get_ingress_address(self):
+        from kubernetes import watch
+
+        w = watch.Watch()
+        for event in w.stream(
+            func=self._networking_api.list_namespaced_ingress, namespace=self._namespace
+        ):
+            if (
+                event["object"].kind == "Ingress"
+                and event["object"].status.load_balancer.ingress is not None
+                and len(event["object"].status.load_balancer.ingress) > 0
+            ):
+                ingress = event["object"].status.load_balancer.ingress[0]
+                ip = ingress.ip
+                host = ingress.hostname if hasattr(ingress, "hostname") else None
+                address = ip or host
+                logger.debug(f"Ingress ip: {ip}, host: {host}, address: {address}")
+                if address is not None:
+                    return f"http://{address}:80"
+
+    def _wait_web_ready(self):
+        loop = new_isolation().loop
+
+        async def get_supervisors():
+            start_time = time.time()
+            while True:
+                try:
+                    cluster_api = WebClusterAPI(self._external_web_endpoint)
```

Comment:
When #110 is merged, we can consider to use Xorbits API instead of mars.

```
from xorbits.core.api import API

api = API.create()
api.list_supervisors()  # we need to add a new one, now only have `list_workers`.
```

Reply:
Add TODO for further enh.